lib/plugins/package/lib/packageService.js

```diff
@@ -18,7 +18,8 @@ module.exports = {
   ],
 
   getIncludes(include) {
-    const packageIncludes = this.serverless.service.package.include || [];
+    const packageIncludes =
+      this.serverless.service.package.patterns || this.serverless.service.package.include || [];
```

Comment:
I think we need to support both `include` and `patterns`, in a way I proposed in this comment: https://github.com/serverless/serverless/pull/8581#discussion_r538467304

Technically we stand between choice:
1. If user provided `patterns`, throw if any `include` and `exclude` configurations are found
2. Support all three

I think it's easier (and not harmful) to support all three

Reply:
You also suggested to crash if both are used in comment: https://github.com/serverless/serverless/issues/8093#issuecomment-676003542 😅

The problem of supporting both `include` and `patterns` is that the order in which they are appended becomes non-evident to the user and can cause tickets simply because the user expects a different order.

The current implementation support using `include/exclude` at the service level and `patterns` at the function level or viceversa. It errors if you use `include/exclude` and `patterns` in the same package object (done during the initialization validation).

I think that it's safer to avoid user confusion.

Do you still think that we should allow using both at the same time?

Reply:
> You also suggested to crash if both are used in comment: #8093 (comment) 😅

That's true. Sorry for confusion. At that point it was not deeply thought. Thing is that crashing is more obtrusive.

Firstly the easy approach to this is to upfront investigate all `package` configs (service-level, function-level, layer-level) and crash if we find that there's `patterns` somewhere and `includes` or `excludes` elsewhere`, and that's aggressive if e.g. user is switching gradually, e.g. refactor first function one, check how it works and then refactor others. It'll be nice to not break that.

Now trying to validate case by case, is confusing, as e.g. for function 1 config (which uses just `include` and `exclude`) may be valid and for function 2 (which uses just `patterns`) may be not (when on service level `include` or `exclude` is used)

I assume it's not harmful to support all (user will always be approached with deprecation notice anyway), and implementation wise it's simplest approach.

> The problem of supporting both include and patterns is that the order in which they are appended becomes non-evident to the user and can cause tickets simply because the user expects a different order.

User should not reason about order in such case, but simply ensure that just `patterns` is used , also again _deprecation_ notice will signal that config stands on quirky ground

Reply:
> Now trying to validate case by case, is confusing, as e.g. for function 1 config (which uses just include and exclude) may be valid and for function 2 (which uses just patterns) may be not (when on service level include or exclude is used)

That's not how it's implemented in this final version of the implementation. Function 1 and 2 will be valid in that example.

Function 3 which uses `patterns` and `exclude` under the same package object would be invalid.
I.e, to get the error you need something like:
```yaml
package:
  patterns:
    - my-file.js
  include:
    - my-other-file.js
```

I think that's reasonable. However, if you still disagree, let me know and I'll just remove it. I'm running out of the time I've allocated for this 😅

Reply:
> Function 1 and 2 will be valid in that example.

I think if we invalidate it on function level, we should definitely also not support combining `include` and `exclude` on service level with `patterns` on function level. That would be even more confusing (what's the logical reason for rejecting just one case?)

Anyway have you read the arguments I've put for supporting all without throwing (?) I take it's simplest thing we can do, and really not harmful

Reply:
I'm not sure if we are understanding each other completely or getting a bit tangled up. 😅
The validation for service level is exactly the same at the function level.
Let's give it one more go and then do whatever change it needed 😄

Currently:
* You can use `patterns` OR `include/exclude` at the service level.
* You can use `patterns` OR `include/exclude` at the function level.
* It doesn't matter if you use pattern in `patterns` at service and `include/exclude` at the function
* It doesn't matter if you use pattern in `include/exclude` at service and `patterns` at the function
* etc...

The only thing that fails, is to mix both `patterns` and `include/exclude` in the same config (regardless of whether it is in the service or function level).

Valid:
```yaml
package:
  patterns:
    - my-file.js

functions:
  myFunction:
    package:
      include:
        - my-other-file.js
      exclude:
        - my-exclude-file.js
```

Valid:
```yaml
package:
  include:
    - my-other-file.js
  exclude:
    - my-exclude-file.js

functions:
  myFunction:
    patterns:
      - my-file.js
```

Valid:
```yaml
package:
  include:
    - my-other-file.js
  exclude:
    - my-exclude-file.js

functions:
  myFunction:
    patterns:
      - my-file.js
  myOtherFunction:
    package:
      include:
        - my-other-file.js
      exclude:
        - my-exclude-file.js
```

Throws:
```yaml
package:
  patterns:
    - my-file.js
  include:
    - my-other-file.js
  exclude:
    - my-exclude-file.js

functions:
  myFunction:
    package:
      include:
        - my-other-file.js
      exclude:
        - my-exclude-file.js
```
Throws:
```yaml
package:
  patterns:
    - my-file.js

functions:
  myFunction:
    package:
      patterns:
        - my-file.js
      include:
        - my-other-file.js
      exclude:
        - my-exclude-file.js
```

Just repeating myself, the reason to not allow that merging is that service goes before package regardless of whether it's `patterns` or `include/exclude`. And `exclude` goes before `include`.
However,  if we mix `include/exclude/pattern` in the same config. We need to decide in which order to merge them. 
It's trivial, I know, `exclude -> include -> patterns`, but users might expect otherwise and create confusion.

> User should not reason about order in such case, but simply ensure that just patterns is used , also again deprecation notice will signal that config stands on quirky ground

That's right! That's why I thought that all combinations should be allowed, except for the one that combines both in the same object.

Was this already clear to you?

Reply:
> You can use patterns OR include/exclude at the function level.
> It doesn't matter if you use pattern in patterns at service and include/exclude at the function

In last comment, my point was that above feels confusing. If we already support mixing both `include` and `patterns` (coming from two configuration sources) to package given lambda, why restrict mixing of `include` and `patterns` coming from one configuration source.

> However, if we mix include/exclude/pattern in the same config. We need to decide in which order to merge them.

Yes, I've proposed order in one of the first comments here: https://github.com/serverless/serverless/pull/8581#discussion_r538467304

> It's trivial, I know, exclude -> include -> patterns, but users might expect otherwise and create confusion.

User should not investigate and debate on such order, instead should just use `pattern`. We'd allow mixing both sas temporary mean simply for ease of migration (no need to be too restrictive here)